File: generate-parentheses/generate-parentheses.java
class Solution {
    List<String> result;
    private void backtrack(String cur, int openN, int closeN) {
        if(openN == 0 && closeN == 0) {
            result.add(cur);
            return;
        }

        if(openN > 0) {
            backtrack(cur + "(", openN-1, closeN);
        }
        if(openN < closeN) {
            backtrack(cur + ")", openN, closeN-1);
        }
    }

    public List<String> generateParenthesis(int n) {
        result = new ArrayList<>();
        backtrack("(", n-1, n);  
        return result;  
    }
}

File: generate-parentheses/generate-parentheses.go
func backtrack(cur string, openN int, closeN int) ([]string) {
    if openN == 0 && closeN == 0 {
        // result = append(result, cur)
        return []string{cur}
    }    

    var left, right []string

    if openN > 0 {
        left = backtrack(cur + "(", openN-1, closeN)
    }
    if openN < closeN {
        right = backtrack(cur + ")", openN, closeN-1)
    }
    return append(left, right...)
}

func generateParenthesis(n int) []string {
    return backtrack("(", n-1, n)
}